cluster/src/main/java/org/apache/iotdb/cluster/log/catchup/CatchUpTask.java

```diff
@@ -0,0 +1,169 @@
+
+    LogCatchUpHandler handler = new LogCatchUpHandler();
+    handler.setAppendSucceed(isMatch);
+    handler.setRaftMember(raftMember);
+    handler.setFollower(node);
+    if (raftMember.getHeader() != null) {
+      request.setHeader(raftMember.getHeader());
+    }
+    request.setLeader(raftMember.getThisNode());
+    // not update follower's commitIndex in order to append the log which index is matchIndex twice
+    request.setLeaderCommit(-1);
+    synchronized (raftMember.getLogManager()) {
+      peer.setNextIndex(raftMember.getLogManager().getLastLogIndex());
+      try {
+        logs = raftMember.getLogManager().getEntries(
+            Math.max(raftMember.getLogManager().getFirstIndex(), peer.getMatchIndex() + 1),
+            peer.getNextIndex() + 1);
+      } catch (Exception e) {
+        logger.error("Unexpected error in logManager's getEntries during matchIndexCheck", e);
+      }
+    }
+
+    int index = logs.size() - 1;
+    EmptyContentLog emptyLog = new EmptyContentLog();
+    while (index >= 0) {
+      Log log = logs.get(index);
+      emptyLog.setCurrLogIndex(log.getCurrLogIndex());
+      emptyLog.setCurrLogTerm(log.getCurrLogTerm());
+      synchronized (raftMember.getTerm()) {
+        // make sure this node is still a leader
+        if (raftMember.getCharacter() != NodeCharacter.LEADER) {
+          logger.debug("Leadership is lost when doing a catch-up to {}, aborting", node);
+          break;
+        }
+        request.setTerm(raftMember.getTerm().get());
+      }
+      if (index > 0) {
+        request.setPrevLogIndex(logs.get(index - 1).getCurrLogIndex());
+        request.setPrevLogTerm(logs.get(index - 1).getCurrLogTerm());
+      } else {
+        request.setPrevLogIndex(log.getCurrLogIndex() - 1);
+        try {
+          request.setPrevLogTerm(raftMember.getLogManager().getTerm(log.getCurrLogIndex() - 1));
+        } catch (Exception e) {
+          logger.error("getTerm failed for newly append entries", e);
+        }
+      }
+
+      handler.setLog(log);
+      request.setEntry(emptyLog.serialize());
+
+      synchronized (isMatch) {
+        AsyncClient client = raftMember.connectNode(node);
+        if (client == null) {
+          break;
+        }
+        client.appendEntry(request, handler);
+        raftMember.getLastCatchUpResponseTime().put(node, System.currentTimeMillis());
+        isMatch.wait(RaftServer.connectionTimeoutInMS);
+      }
+      if (isMatch.get()) {
+        // if follower return RESPONSE.AGREE with this empty log, then start sending real logs from this emptyContentLog's index.
+        // which means the log which index is matchIndex will be send twice, but at the first time it sent an empty log.
+        logs.subList(0, index).clear();
+        if (logger.isDebugEnabled()) {
+          logger.debug("{} makes {} catch up with {} cached logs", raftMember.getName(), node,
+              logs.size());
+        }
+        return false;
+      }
+      index--;
```

Comment:
It would be better to perform a binary search, but it is fine for now.

Reply:
Inconsistencies are generally rare, and there aren't many missing logs, so this may actually be enough. If a performance bottleneck occurs, I will modify here ~